rasa/shared/utils/common.py

```diff
@@ -66,6 +68,64 @@ def _lazyprop(self):
     return _lazyprop
 
 
+def cached_method(f: Callable[..., Any]) -> Callable[..., Any]:
+    """Caches method calls based on the call's `args` and `kwargs`.
+
+    Works for `async` and `sync` methods. Don't apply this to functions.
+
+    Returns:
+        The return value which the method gives for the first call with the given
+        arguments.
+    """
+
+    class Cache:
+        """Helper class to abstract the caching details."""
+
+        def __init__(self, caching_object: object, args: Any, kwargs: Any) -> None:
+            self.caching_object = caching_object
+            self.cache = getattr(caching_object, self._cache_name(), {})
+            # noinspection PyUnresolvedReferences
+            self.cache_key = functools._make_key(  # pytype: disable=module-attr
+                args, kwargs, typed=False
+            )
+
+        def _cache_name(self) -> Text:
+            return f"_cached_{self.caching_object.__class__.__name__}_{f.__name__}"
+
+        def is_cached(self) -> bool:
+            return self.cache_key in self.cache
+
+        def cache_result(self, result: Any) -> None:
+            self.cache[self.cache_key] = result
+            setattr(self.caching_object, self._cache_name(), self.cache)
+
+        def cached_result(self) -> Any:
+            return self.cache[self.cache_key]
+
+    if asyncio.iscoroutinefunction(f):
+
+        @functools.wraps(f)
+        async def decorated(self: object, *args: Any, **kwargs: Any) -> Any:
+            cache = Cache(self, args, kwargs)
+            if not cache.is_cached():
+                to_cache = await f(self, *args, **kwargs)
```

Comment:
should we rather cache the future instead of the awaited result? we might call f multiple times if we are still awaiting the return of f but another caller called this function before it completed

Reply:
Done. I wrote a test for this edge case but decided not to add it as it depends on timing and I don't a test in our CI which depends on timing.